src/MLMCManager.cpp

```diff
@@ -0,0 +1,450 @@
+        }
+
+        ss3 << "c_coefficient" << sample_count_[level] << ".vector";
+        std::ofstream out3(ss3.str().c_str());
+        fine_coefficient.Print(out3, 1);
+    }
+    sample_count_[level]++;
+}
+
+double MLMCManager::GetEstimate() const
+{
+    double out = 0.0;
+    for (const auto& v : mean_)
+    {
+        out += v;
+    }
+    return out;
+}
+
+/**
+   This ugly implementation is to use Vector::Swap which does not
+   exist for BlockVector, so we wrap Vector in temporary BlockVector
+   objects that share the data with Vector.
+*/
+mfem::BlockVector InterpolateToFine(const Upscale& upscale, int level,
+                                    const mfem::BlockVector& in)
+{
+    mfem::Vector vec1, vec2;
+    vec1 = in;
+    for (int k = level; k > 0; k--)
+    {
+        MFEM_ASSERT(vec1.Size() == upscale.GetMatrix(k).GetBlockOffsets().Last(),
+                    "Sizes do not work!");
+        mfem::BlockVector block_vec1(vec1.GetData(),
+                                     upscale.GetMatrix(k).GetBlockOffsets());
+        vec2.SetSize(upscale.GetMatrix(k - 1).GetBlockOffsets().Last());
+        mfem::BlockVector block_vec2(vec2.GetData(),
+                                     upscale.GetMatrix(k - 1).GetBlockOffsets());
+        /// Interpolate from k to the finer k-1
+        upscale.Interpolate(k, block_vec1, block_vec2);
+        vec2.Swap(vec1);
+    }
+    mfem::BlockVector out(upscale.GetMatrix(0).GetBlockOffsets());
+    MFEM_ASSERT(out.Size() == vec1.Size(), "Sizes do not work!");
+    // ((mfem::Vector) out) = vec1; // doesn't work for some reason (valgrind doesn't like it)
+    for (int i = 0; i < out.Size(); ++i)
+    {
+        out[i] = vec1[i];
+    }
+    return out;
+}
+
+/// restrict from finest level (level 0) to given level
+mfem::BlockVector RestrictToLevel(const Upscale& upscale,
+                                  int level, const mfem::BlockVector& in)
+{
+    mfem::Vector vec1, vec2;
+    vec1 = in;
+    for (int k = 0; k < level; ++k)
+    {
+        MFEM_ASSERT(vec1.Size() == upscale.GetMatrix(k).GetBlockOffsets().Last(),
+                    "Sizes do not work!");
+        mfem::BlockVector block_vec1(vec1.GetData(),
+                                     upscale.GetMatrix(k).GetBlockOffsets());
+        vec2.SetSize(upscale.GetMatrix(k + 1).GetBlockOffsets().Last());
+        vec2 = 0.0; // ????
+        mfem::BlockVector block_vec2(vec2.GetData(),
+                                     upscale.GetMatrix(k + 1).GetBlockOffsets());
+        upscale.Restrict(k + 1, block_vec1, block_vec2);
+        vec2.Swap(vec1);
+    }
+    mfem::BlockVector out(upscale.GetMatrix(level).GetBlockOffsets());
+    MFEM_ASSERT(out.Size() == vec1.Size(), "Sizes do not work!");
+    // ((mfem::Vector) out) = vec1;
+    for (int i = 0; i < out.Size(); ++i)
+    {
+        out[i] = vec1[i];
+    }
+    return out;
+}
+
+void MLMCManager::CoarseSample(bool verbose)
```

Comment:
This shares a lot of code with `MLMCManager::FineSample`. 

Maybe we can think of `MLMCManager` and `Upscale` as two different objects in that `Upscale` provides a "M-level" hierarchy, `MLMCManager` can do a "m-level" MLMC simulation for any "m" <= "M". This way, at the coarsest level "m-1" we will always just call this function (in the case of fine level MC, `num_levels_` = 0, so this function will pretty much do what `FineSample` does, except the rhs, which is commented below).

For any level "k" < "m-1", the CorrectionSample will be call.